src/OpenTelemetry.Exporter.Prometheus.HttpListener/Internal/PrometheusMetric.cs

```diff
@@ -0,0 +1,293 @@
+            // type-specific suffixes.
+            // https://github.com/open-telemetry/opentelemetry-specification/blob/b2f923fb1650dde1f061507908b834035506a796/specification/compatibility/prometheus_and_openmetrics.md#L242-L246
+            if (!sanitizedName.Contains(sanitizedUnit))
+            {
+                sanitizedName = sanitizedName + "_" + sanitizedUnit;
+            }
+        }
+
+        // If the metric name for monotonic Sum metric points does not end in a suffix of `_total` a suffix of `_total` MUST be added by default, otherwise the name MUST remain unchanged.
+        // Exporters SHOULD provide a configuration option to disable the addition of `_total` suffixes.
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/b2f923fb1650dde1f061507908b834035506a796/specification/compatibility/prometheus_and_openmetrics.md#L286
+        if (type == PrometheusType.Counter && !sanitizedName.EndsWith("_total"))
+        {
+            sanitizedName += "_total";
+        }
+
+        // Special case: Converting "1" to "ratio".
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/b2f923fb1650dde1f061507908b834035506a796/specification/compatibility/prometheus_and_openmetrics.md#L239
+        if (type == PrometheusType.Gauge && unit == "1" && !sanitizedName.Contains("ratio"))
+        {
+            sanitizedName += "_ratio";
+        }
+
+        this.Name = sanitizedName;
+        this.Unit = sanitizedUnit;
+        this.Type = type;
+    }
+
+    public string Name { get; }
+
+    public string Unit { get; }
+
+    public PrometheusType Type { get; }
+
+    public static PrometheusMetric Create(Metric metric)
+    {
+        return new PrometheusMetric(metric.Name, metric.Unit, GetPrometheusType(metric));
+    }
+
+    internal static string SanitizeMetricName(string metricName)
+    {
+        StringBuilder sb = null;
+        var lastCharUnderscore = false;
+
+        for (var i = 0; i < metricName.Length; i++)
+        {
+            var c = metricName[i];
+
+            if (i == 0 && char.IsNumber(c))
+            {
+                sb ??= CreateStringBuilder(metricName);
+                sb.Append('_');
+                lastCharUnderscore = true;
+                continue;
+            }
+
+            if (!char.IsLetterOrDigit(c) && c != ':')
+            {
+                if (!lastCharUnderscore)
+                {
+                    lastCharUnderscore = true;
+                    sb ??= CreateStringBuilder(metricName);
+                    sb.Append('_');
+                }
+            }
+            else
+            {
+                sb ??= CreateStringBuilder(metricName);
+                sb.Append(c);
+                lastCharUnderscore = false;
+            }
+        }
+
+        return sb?.ToString() ?? metricName;
+
+        static StringBuilder CreateStringBuilder(string name) => new StringBuilder(name.Length);
```

Comment:
You could probably use `stackalloc` for this method and avoid the `StringBuilder` allocation.

[Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax:~:text=They%20can%20have%20a%20maximum%20length%20of%2063%20characters.) doesn't allow more than 63 characters for the metric name. Spec has the same restriction for Unit but we haven't enforced that in the SDK yet.

Reply:
It looks like the spec recently got updated to allow for a higher number of characters: https://github.com/open-telemetry/opentelemetry-specification/issues/3422

Considering, this might change again in future, it's better to avoid assuming anything about the metric name length.